app/questionnaire/navigation.py

```diff
@@ -31,110 +31,143 @@ def build_navigation(self, current_group_id, current_group_instance):
         :param current_group_instance:
         :return:
         """
-        if self.survey_json.get('navigation', False) is False:
+
+        navigation_block = self.survey_json.get('navigation')
+        if navigation_block is None or navigation_block.get("visible", True) is False:
             return None
 
         navigation = []
 
-        last_group_id = SchemaHelper.get_last_group_id(self.survey_json)
+        for section in navigation_block['sections']:
+            non_skipped_groups = self._get_non_skipped_groups(section)
+            if len(non_skipped_groups) == 0:
+                continue
 
-        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
-            logger.debug("building frontend navigation", group_id=group['id'])
+            first_group_id = non_skipped_groups[0]
+            first_block_id = SchemaHelper.get_first_block_id_for_group(self.survey_json, first_group_id)
 
-            first_location = Location(group['id'], 0, group['blocks'][0]['id'])
-            last_block_in_group = SchemaHelper.get_last_block_in_group(group)
+            last_group_block_id = self._get_last_group_and_block_id(non_skipped_groups)
```

Comment:
I think it would be clearer if we replaced last with completed, so we'd have `completion_group_and_block_id`, `_get_completion_group_and_block_id`, and further on `completion_group_id` and `completion_block_id`